DataFormats/TrackerRecHit2D/interface/SiPixelRecHitQuality.h

```diff
@@ -110,7 +109,13 @@ class SiPixelRecHitQuality {
     //------------------------------------------------------
     //
     inline void setProbabilityXY(float prob, QualWordType& qualWord) const {
-      if (prob < 0 || prob > 1) {
+      if (prob < 0.0f) {
+        warningOutOfBoundProb("XY", prob, qualWord);
+        prob = 0;
+      } else if (prob > 1.0f && prob < 1.0f + 32 * std::numeric_limits<float>::epsilon()) {
+        warningOutOfBoundProbDueToPrecision("XY", prob, qualWord);
+        prob = 1;
+      } else if (prob > 1.0f + 32 * std::numeric_limits<float>::epsilon()) {
```

Comment:
the range check has now grown and is repeated twice. It's time to refactor into a separate method, e.g. `float checkProbRange(...`

Reply:
Sure, I can do that. Can we agree that the  multiple of 32 is fine? I can happily make it the meaning of life too, if that's prefered :)

Reply:
from the description in https://github.com/cms-sw/cmssw/pull/36120#discussion_r752558858, I do not see why 32 would be a valid value.

Now that I know where the origin is, why can't we clamp the result?
in RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco.cc
instead of `    probx = 1. - TMath::Gamma(hndof, hchi2);` use 
`probx = std::clamp(1. - TMath::Gamma(hndof, hchi2), 0., 1.);`

Reply:
It's not actually coming from the probX and probY but from the probQ. Anyway about probXY, look here
https://github.com/cms-sw/cmssw/blob/master/RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco.cc#L1157-L1164
"Don't return exact zeros for the probability"

and then for probQ
Sometimes that is meant to be a negative number
https://github.com/cms-sw/cmssw/blob/master/RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco.cc#L1198

Reply:
> Sure, I can do that. Can we agree that the multiple of 32 is fine? I can happily make it the meaning of life too, if that's prefered :)

I am now confused by the changes in the PR. The earliest variant as e.g. in 07c0b5d481d7c6ff2bb718f6c056160d6925a0c9 has extended the range check and was not sending a warning for 1+epsilon.
The current version seems to only change the warning text (and the "clamped" value), but would still issue a warning for the same values as in the baseline.

I feel somewhat confident about silencing the warning for `1+(1 or 2 )* epsilon`.

For larger deviations, the current choice of messaging seems OK.